Computed display of form inside table elements should be display:none

Ahmad-S792 opened the pull request:
#### 78b54052c50b05e6861860cc08ba9915d8c3404f
<pre>
Computed display of form inside table elements should be display:none

<a href="https://bugs.webkit.org/show_bug.cgi?id=252541">https://bugs.webkit.org/show_bug.cgi?id=252541</a>
rdar://problem/105932450

Reviewed by Tim Nguyen.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox and web-specification.

Web-Spec: <a href="https://html.spec.whatwg.org/multipage/rendering.html#tables-2">https://html.spec.whatwg.org/multipage/rendering.html#tables-2</a>

The patch adds UA stylesheet rule as expected from the web-specification to change display
type of form inside table elements to &quot;display: none&quot; by introducing internal pseudo class &apos;:-internal-html-document&apos;.

* Source/WebCore/css/html.css: Add new UA rules for form in tables
* Source/WebCore/css/CSSSelector.h: New Pseudo class &apos;PseudoClassHtmlDocument&apos;
* Source/WebCore/css/CSSSelector.cpp:
(CSSSelector::selectorText): As above to translate in &apos;:-internal-html-document&apos;
* Source/WebCore/css/SelectorChecker.cpp:
(SelectorChecker::checkOne): Add case for &apos;PseudoClassHtmlDocument&apos;
* Source/WebCore/css/SelectorCheckerTestFunctions.h: New bool function &apos;matchesHtmlDocumentPseudoClass&apos; to return &apos;HTML&apos; document
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in: Add &apos;-internal-html-document&apos; in list
* Source/WebCore/cssjit/SelectorCompiler.cpp: Add in namespace similar to other
(operationMatchesIsHtmlPseudoClass): New function to return &apos;matchesHtmlDocumentPseudoClass&apos;
(addPseudoClassType): Add new case &apos;PseudoClassHtmlDocument&apos;
* Source/WebCore/html/HTMLFormElement.cpp:
(HTMLFormElement::rendererIsNeeded): Deleted hardcoded logic of forms in table layout
(copyNonAttributePropertiesFromElement): Deleted
* Source/WebCore/html/HTMLFormElement.h: Deleted unused definition after above changes
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(CSSSelectorParser::consumePseudo): Restrict exposing internal pseudo class to web
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(HTMLConstructionSite::insertHTMLFormElement): Remove &apos;isDemoted&apos; and usage
* Source/WebCore/html/parser/HTMLConstructionSite.h: Remove &apos;isDemoted&apos; from definition
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(HTMLTreeBuilder::processStartTagForInTable): Update since &apos;isDemoted&apos; is now gone
* LayoutTests/fast/css/pseudo-class-internal.html: Update for new internal pseudo class
* LayoutTests/fast/css/pseudo-class-internal-expected.txt&quot; Update Expectations for above
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/form-in-tables-expected.txt: Rebaselined

Canonical link: <a href="https://commits.webkit.org/265283@main">https://commits.webkit.org/265283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59840030d0cd2f621b9445f5a5849878867676a0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10491 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12131 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10061 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10503 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13008 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8823 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12530 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9479 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16739 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9929 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12880 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10083 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8171 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9242 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/9373 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2517 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13499 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9945 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/parser/HTMLConstructionSite.cpp:
```diff
@@ -517,7 +517,8 @@ void HTMLConstructionSite::insertHTMLFormElement(AtomHTMLToken&& token, bool isD
         m_form = &formElement;
-    formElement.setDemoted(isDemoted);
+    if (isDemoted)
+        ownerDocumentForCurrentNode();
```
We can just remove this code. It seems like the Chromium is just using the document for an use counter, which we don't have.

We should also remove the `isDemoted` parameter from this function and its callsites.

Ahmad-S792 replied:
Done.

nt1m commented on Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
```diff
@@ -1,4 +1,5 @@
 -khtml-drag
+-internal-is-html
```
Can we rename this and related C++ names to -internal-is-html-document ? 

:-internal-is-html can be confusing, since it's unclear if it's the element or the document that is HTML.

Ahmad-S792 replied:
Just this or also renaming 'matchesIsHtmlPseudoClass' as 'matchesIsHtmlDocumentPseudoClass' etc. Happy to do this since now it works and this is just renaming.

nt1m replied:
Also renaming 'matchesIsHtmlPseudoClass' as 'matchesIsHtmlDocumentPseudoClass'

Ahmad-S792 replied:
On it. :-)

nt1m replied:
Actually, can you also remove the "Is" part? since the it's not the element that's the html document

nt1m replied:
Same from the pseudo class name: `:-internal-html-document`

Ahmad-S792 replied:
Ops! Just pushed with 'is' part. Let me do another quick one. :-)

Ahmad-S792 replied:
[5984003](https://github.com/WebKit/WebKit/commit/59840030d0cd2f621b9445f5a5849878867676a0) now with all asked changes and renames.

Ahmad-S792 replied:
Commit log also updated. Marking this as resolved.

Ahmad-S792 replied:
@nt1m - now renamed to 'internal-html-document' and all other related C++ parts as well.

nt1m reviewed (approved):
Looks great, thanks! let's wait for EWS results and merge

Ahmad-S792 committed:
Computed display of form inside table elements should be display:none

https://bugs.webkit.org/show_bug.cgi?id=252541
rdar://problem/105932450

Reviewed by Tim Nguyen.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox and web-specification.

Web-Spec: https://html.spec.whatwg.org/multipage/rendering.html#tables-2

The patch adds UA stylesheet rule as expected from the web-specification to change display
type of form inside table elements to "display: none" by introducing internal pseudo class ':-internal-html-document'.

* Source/WebCore/css/html.css: Add new UA rules for form in tables
* Source/WebCore/css/CSSSelector.h: New Pseudo class 'PseudoClassHtmlDocument'
* Source/WebCore/css/CSSSelector.cpp:
(CSSSelector::selectorText): As above to translate in ':-internal-html-document'
* Source/WebCore/css/SelectorChecker.cpp:
(SelectorChecker::checkOne): Add case for 'PseudoClassHtmlDocument'
* Source/WebCore/css/SelectorCheckerTestFunctions.h: New bool function 'matchesHtmlDocumentPseudoClass' to return 'HTML' document
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in: Add '-internal-html-document' in list
* Source/WebCore/cssjit/SelectorCompiler.cpp: Add in namespace similar to other
(operationMatchesIsHtmlPseudoClass): New function to return 'matchesHtmlDocumentPseudoClass'
(addPseudoClassType): Add new case 'PseudoClassHtmlDocument'
* Source/WebCore/html/HTMLFormElement.cpp:
(HTMLFormElement::rendererIsNeeded): Deleted hardcoded logic of forms in table layout
(copyNonAttributePropertiesFromElement): Deleted
* Source/WebCore/html/HTMLFormElement.h: Deleted unused definition after above changes
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(CSSSelectorParser::consumePseudo): Restrict exposing internal pseudo class to web
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(HTMLConstructionSite::insertHTMLFormElement): Remove 'isDemoted' and usage
* Source/WebCore/html/parser/HTMLConstructionSite.h: Remove 'isDemoted' from definition
* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(HTMLTreeBuilder::processStartTagForInTable): Update since 'isDemoted' is now gone
* LayoutTests/fast/css/pseudo-class-internal.html: Update for new internal pseudo class
* LayoutTests/fast/css/pseudo-class-internal-expected.txt" Update Expectations for above
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/tables/form-in-tables-expected.txt: Rebaselined

Canonical link: https://commits.webkit.org/265283@main
